
<!DOCTYPE html>
<!--[if IE 9]><html class="lt-ie10" lang="en" > <![endif]-->
<html class="no-js" lang="en" >
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Chapter 6. Exceptions and File Handling</title>
        <link rev="made" href="josh@servercobra.com" />
        <meta name="generator" content="convert_to_jinja2.py" />
        <meta name="keywords" content="Python, Dive Into Python, tutorial, object-oriented, programming, documentation, book, free" />
        <meta name="description" content="Python from novice to pro" />
        <link rel="home" href="http://www.diveintopython.net/" title="Dive Into Python" />
        
    
    <link rel="up" href="../toc/index.html" title=" Dive Into Python" />
    
    
    
    <link rel="next" href="file_objects.html" title=" 6.2.Working with File Objects" />
    

        <link rel="stylesheet" href="../css/normalize.css">
        <link rel="stylesheet" href="../css/foundation.css">
        <link rel="stylesheet" href="../css/diveintopython.css" type="text/css" />
        <script src="../js/modernizr.js"></script>
    </head>
    <body>
        <div class="row">
            <div class="small-12 columns" id="Header">
                <div class="row">
                    <div class="small-11 columns">
                        
    
    <td align="left" colspan="5" id="breadcrumb" valign="top">You are here: <a href="../index.html">Home</a>&#160;&gt;&#160;<a href="../toc/index.html">Dive Into Python</a>&#160;&gt;&#160;<span class="thispage">Exceptions and File Handling</span></td>
    

                    </div>
                    <div class="small-1 columns">
                        
    <td align="right" id="navigation" valign="top">&#160;&#160;&#160;<a href="../object_oriented_framework/summary.html" title="Prev: &#8220;Summary&#8221;">&lt;&lt;</a>&#160;&#160;&#160;<a href="file_objects.html" title="Next: &#8220;Working with File Objects&#8221;">&gt;&gt;</a></td>

                    </div>
                </div>

                <div id="logocontainer">
                    <h1 id="logo"><a href="http://www.diveintopython.net/index.html" accesskey="1">Dive Into Python</a></h1>
                    <p id="tagline">Python from novice to pro</p>
                    <a href="http://www.amazon.com/gp/product/B001GIOFN8/ref=as_li_tf_tl?ie=UTF8&tag=diveintopython-20&linkCode=as2&camp=1789&creative=9325&creativeASIN=B001GIOFN8"><img src="../images/amazon-buy-button.gif"></a>
                </div>
            </div>
        </div>
        <div class="row">
            <div class="small-12 columns">
                
    <div class="titlepage">
<div>
<div>
<h2 class="title"><a name="filehandling"></a>Chapter&#160;6.&#160;Exceptions and File Handling
                  </h2>
</div>
</div>
<div></div>
</div>

            </div>
        </div>
        <div class="row">
            <div class="small-12 columns">
                
    <div class="toc">
<ul>
<li><span class="section"><a href="index.html#fileinfo.exception">6.1. Handling Exceptions</a></span><ul>
<li><span class="section"><a href="index.html#d0e14344">6.1.1. Using Exceptions For Other Purposes</a></span></li>
</ul>
</li>
<li><span class="section"><a href="file_objects.html">6.2. Working with File Objects</a></span><ul>
<li><span class="section"><a href="file_objects.html#d0e14670">6.2.1. Reading Files</a></span></li>
<li><span class="section"><a href="file_objects.html#d0e14800">6.2.2. Closing Files</a></span></li>
<li><span class="section"><a href="file_objects.html#d0e14928">6.2.3. Handling I/O Errors</a></span></li>
<li><span class="section"><a href="file_objects.html#d0e15055">6.2.4. Writing to Files</a></span></li>
</ul>
</li>
<li><span class="section"><a href="for_loops.html">6.3. Iterating with for Loops</a></span></li>
<li><span class="section"><a href="more_on_modules.html">6.4. Using sys.modules</a></span></li>
<li><span class="section"><a href="os_module.html">6.5. Working with Directories</a></span></li>
<li><span class="section"><a href="all_together.html">6.6. Putting It All Together</a></span></li>
<li><span class="section"><a href="summary.html">6.7. Summary</a></span></li>
</ul>
</div>

            </div>
        </div>
        <div class="row">
            <div class="small-12 columns">
                
    <div class="abstract">
<p>In this chapter, you will dive into exceptions, file objects, <tt class="literal">for</tt> loops, and the <tt class="filename">os</tt> and <tt class="filename">sys</tt> modules.  If you've used exceptions in another programming language, you can skim the first section to get a sense of <span class="application">Python</span>'s syntax.  Be sure to tune in again for file handling.
            </p>
</div>

            </div>
        </div>
        <div class="row">
            <div class="small-12 columns">
                
    <div class="section" lang="en">
<div class="titlepage">
<div>
<div>
<h2 class="title"><a name="fileinfo.exception"></a>6.1.&#160;Handling Exceptions
                     </h2>
</div>
</div>
<div></div>
</div>
<div class="toc">
<ul>
<li><span class="section"><a href="index.html#d0e14344">6.1.1. Using Exceptions For Other Purposes</a></span></li>
</ul>
</div>
<div class="abstract">
<p>Like many other programming languages, <span class="application">Python</span> has exception handling via <tt class="literal">try...except</tt> blocks.
               </p>
</div><a name="compare.exceptions.java"></a><table border="0" class="note" summary="">
<tr>
<td align="center" rowspan="2" valign="top" width="1%"><img alt="Note" height="24" src="../images/note.png" title="" width="24"/></td>
</tr>
<tr>
<td align="left" colspan="2" valign="top" width="99%"><span class="application">Python</span> uses <tt class="literal">try...except</tt> to handle exceptions and <tt class="literal">raise</tt> to generate them.  <span class="application">Java</span> and <span class="application"><span class="acronym">C++</span></span> use <tt class="literal">try...catch</tt> to handle exceptions, and <tt class="literal">throw</tt> to generate them.
                  </td>
</tr>
</table>
<p>Exceptions are everywhere in <span class="application">Python</span>.  Virtually every module in the standard <span class="application">Python</span> library uses them, and <span class="application">Python</span> itself will raise them in a lot of different circumstances.  You've already seen them repeatedly throughout this book.
            </p>
<div class="itemizedlist">
<ul>
<li><a href="../native_data_types/index.html#odbchelper.dict.define" title="Example&#160;3.1.&#160;Defining a Dictionary">Accessing a non-existent dictionary key</a> will raise a <tt class="errorcode">KeyError</tt> exception.
                  </li>
<li><a href="../native_data_types/lists.html#odbchelper.list.search" title="Example&#160;3.12.&#160;Searching a List">Searching a list for a non-existent value</a> will raise a <tt class="errorcode">ValueError</tt> exception.
                  </li>
<li><a href="../native_data_types/tuples.html#odbchelper.tuplemethods" title="Example&#160;3.16.&#160;Tuples Have No Methods">Calling a non-existent method</a> will raise an <tt class="errorcode">AttributeError</tt> exception.
                  </li>
<li><a href="../native_data_types/declaring_variables.html#odbchelper.unboundvariable" title="Example&#160;3.18.&#160;Referencing an Unbound Variable">Referencing a non-existent variable</a> will raise a <tt class="errorcode">NameError</tt> exception.
                  </li>
<li><a href="../native_data_types/formatting_strings.html#odbchelper.stringformatting.coerce" title="Example&#160;3.22.&#160;String Formatting vs. Concatenating">Mixing datatypes without coercion</a> will raise a <tt class="errorcode">TypeError</tt> exception.
                  </li>
</ul>
</div>
<p>In each of these cases, you were simply playing around in the <span class="application">Python</span> <span class="acronym">IDE</span>: an error occurred, the exception was printed (depending on your <span class="acronym">IDE</span>, perhaps in an intentionally jarring shade of red), and that was that.  This is called an <span class="emphasis"><em>unhandled</em></span> exception.  When the exception was raised, there was no code to explicitly notice it and deal with it, so it bubbled its
               way back to the default behavior built in to <span class="application">Python</span>, which is to spit out some debugging information and give up.  In the <span class="acronym">IDE</span>, that's no big deal, but if that happened while your actual <span class="application">Python</span> program was running, the entire program would come to a screeching halt.
            </p>
<p>An exception doesn't need result in a complete program crash, though.  Exceptions, when raised, can be <span class="emphasis"><em>handled</em></span>.  Sometimes an exception is really because you have a bug in your code (like accessing a variable that doesn't exist), but
               many times, an exception is something you can anticipate.  If you're opening a file, it might not exist.  If you're connecting
               to a database, it might be unavailable, or you might not have the correct security credentials to access it.  If you know
               a line of code may raise an exception, you should handle the exception using a <tt class="literal">try...except</tt> block.
            </p>
<div class="example"><a name="d0e14243"></a><h3 class="title">Example&#160;6.1.&#160;Opening a Non-Existent File</h3><pre class="screen"><tt class="prompt">&gt;&gt;&gt; </tt><span class="userinput">fsock = open(<span class="pystring">"/notthere"</span>, <span class="pystring">"r"</span>)</span>      <a name="fileinfo.exceptions.1.1"></a><img alt="1" border="0" height="12" src="../images/callouts/1.png" width="12">
<span class="traceback">Traceback (innermost last):
  File "&lt;interactive input&gt;", line 1, in ?
IOError: [Errno 2] No such file or directory: '/notthere'</span>
<tt class="prompt">&gt;&gt;&gt; </tt><span class="userinput"><span class="pykeyword">try</span>:</span>
<tt class="prompt">...     </tt><span class="userinput">fsock = open(<span class="pystring">"/notthere"</span>)</span>       <a name="fileinfo.exceptions.1.2"></a><img alt="2" border="0" height="12" src="../images/callouts/2.png" width="12">
<tt class="prompt">... </tt><span class="userinput"><span class="pykeyword">except</span> IOError:</span>                     <a name="fileinfo.exceptions.1.3"></a><img alt="3" border="0" height="12" src="../images/callouts/3.png" width="12">
<tt class="prompt">...     </tt><span class="userinput"><span class="pykeyword">print</span> <span class="pystring">"The file does not exist, exiting gracefully"</span></span>
<tt class="prompt">... </tt><span class="userinput"><span class="pykeyword">print</span> <span class="pystring">"This line will always print"</span></span> <a name="fileinfo.exceptions.1.4"></a><img alt="4" border="0" height="12" src="../images/callouts/4.png" width="12">
<span class="computeroutput">The file does not exist, exiting gracefully
This line will always print</span></img></img></img></img></pre><div class="calloutlist">
<table border="0" summary="Callout list">
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.1.1"><img alt="1" border="0" height="12" src="../images/callouts/1.png" width="12"/></a>
</td>
<td align="left" valign="top">Using the built-in <tt class="function">open</tt> function, you can try to open a file for reading (more on <tt class="function">open</tt> in the next section).  But the file doesn't exist, so this raises the <tt class="errorcode">IOError</tt> exception.  Since you haven't provided any explicit check for an <tt class="errorcode">IOError</tt> exception, <span class="application">Python</span> just prints out some debugging information about what happened and then gives up.
                        </td>
</tr>
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.1.2"><img alt="2" border="0" height="12" src="../images/callouts/2.png" width="12"/></a>
</td>
<td align="left" valign="top">You're trying to open the same non-existent file, but this time you're doing it within a <tt class="literal">try...except</tt> block.
                        </td>
</tr>
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.1.3"><img alt="3" border="0" height="12" src="../images/callouts/3.png" width="12"/></a>
</td>
<td align="left" valign="top">When the <tt class="function">open</tt> method raises an <tt class="errorcode">IOError</tt> exception, you're ready for it.  The <tt class="literal">except IOError:</tt> line catches the exception and executes your own block of code, which in this case just prints a more pleasant error message.
                        </td>
</tr>
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.1.4"><img alt="4" border="0" height="12" src="../images/callouts/4.png" width="12"/></a>
</td>
<td align="left" valign="top">Once an exception has been handled, processing continues normally on the first line after the <tt class="literal">try...except</tt> block.  Note that this line will always print, whether or not an exception occurs.  If you really did have a file called
                           <tt class="filename">notthere</tt> in your root directory, the call to <tt class="function">open</tt> would succeed, the <tt class="literal">except</tt> clause would be ignored, and this line would still be executed.
                        </td>
</tr>
</table>
</div>
</div>
<p>Exceptions may seem unfriendly (after all, if you don't catch the exception, your entire program will crash), but consider
               the alternative.  Would you rather get back an unusable file object to a non-existent file?  You'd need to check its validity
               somehow anyway, and if you forgot, somewhere down the line, your program would give you strange errors somewhere down the
               line that you would need to trace back to the source.  I'm sure you've experienced this, and you know it's not fun.  With
               exceptions, errors occur immediately, and you can handle them in a standard way at the source of the problem.
            </p>
<div class="section" lang="en">
<div class="titlepage">
<div>
<div>
<h3 class="title"><a name="d0e14344"></a>6.1.1.&#160;Using Exceptions For Other Purposes
                        </h3>
</div>
</div>
<div></div>
</div>
<p>There are a lot of other uses for exceptions besides handling actual error conditions.  A common use in the standard <span class="application">Python</span> library is to try to import a module, and then check whether it worked.  Importing a module that does not exist will raise
                  an <tt class="errorcode">ImportError</tt> exception.  You can use this to define multiple levels of functionality based on which modules are available at run-time,
                  or to support multiple platforms (where platform-specific code is separated into different modules).
               </p>
<p>You can also define your own exceptions by creating a class that inherits from the built-in <tt class="classname">Exception</tt> class, and then raise your exceptions with the <tt class="function">raise</tt> command.  See the further reading section if you're interested in doing this.
               </p>
<p>The next example demonstrates how to use an exception to support platform-specific functionality.  This code comes from the
                  <tt class="filename">getpass</tt> module, a wrapper module for getting a password from the user.  Getting a password is accomplished differently on <span class="acronym">UNIX</span>, Windows, and <span class="abbrev">Mac</span> <span class="acronym">OS</span> platforms, but this code encapsulates all of those differences.
               </p>
<div class="example"><a name="crossplatform.example"></a><h3 class="title">Example&#160;6.2.&#160;Supporting Platform-Specific Functionality</h3><pre class="programlisting">
  <span class="pycomment"># Bind the name getpass to the appropriate function</span>
  <span class="pykeyword">try</span>:
      <span class="pykeyword">import</span> termios, TERMIOS                     <a name="fileinfo.exceptions.2.1"></a><img alt="1" border="0" height="12" src="../images/callouts/1.png" width="12">
  <span class="pykeyword">except</span> ImportError:
      <span class="pykeyword">try</span>:
          <span class="pykeyword">import</span> msvcrt                           <a name="fileinfo.exceptions.2.2"></a><img alt="2" border="0" height="12" src="../images/callouts/2.png" width="12">
      <span class="pykeyword">except</span> ImportError:
          <span class="pykeyword">try</span>:
              <span class="pykeyword">from</span> EasyDialogs <span class="pykeyword">import</span> AskPassword <a name="fileinfo.exceptions.2.3"></a><img alt="3" border="0" height="12" src="../images/callouts/3.png" width="12">
          <span class="pykeyword">except</span> ImportError:
              getpass = default_getpass           <a name="fileinfo.exceptions.2.4"></a><img alt="4" border="0" height="12" src="../images/callouts/4.png" width="12">
          <span class="pykeyword">else</span>:                                   <a name="fileinfo.exceptions.2.5"></a><img alt="5" border="0" height="12" src="../images/callouts/5.png" width="12">
              getpass = AskPassword
      <span class="pykeyword">else</span>:
          getpass = win_getpass
  <span class="pykeyword">else</span>:
      getpass = unix_getpass</img></img></img></img></img></pre><div class="calloutlist">
<table border="0" summary="Callout list">
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.2.1"><img alt="1" border="0" height="12" src="../images/callouts/1.png" width="12"/></a>
</td>
<td align="left" valign="top"><tt class="filename">termios</tt> is a <span class="acronym">UNIX</span>-specific module that provides low-level control over the input terminal.  If this module is not available (because it's not
                              on your system, or your system doesn't support it), the import fails and <span class="application">Python</span> raises an <tt class="errorcode">ImportError</tt>, which you catch.
                           </td>
</tr>
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.2.2"><img alt="2" border="0" height="12" src="../images/callouts/2.png" width="12"/></a>
</td>
<td align="left" valign="top">OK, you didn't have <tt class="filename">termios</tt>, so let's try <tt class="filename">msvcrt</tt>, which is a Windows-specific module that provides an <span class="acronym">API</span> to many useful functions in the Microsoft <span class="application">Visual C++</span> runtime services.  If this import fails, <span class="application">Python</span> will raise an <tt class="errorcode">ImportError</tt>, which you catch.
                           </td>
</tr>
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.2.3"><img alt="3" border="0" height="12" src="../images/callouts/3.png" width="12"/></a>
</td>
<td align="left" valign="top">If the first two didn't work, you try to import a function from <tt class="filename">EasyDialogs</tt>, which is a <span class="abbrev">Mac</span> <span class="acronym">OS</span>-specific module that provides functions to pop up dialog boxes of various types.  Once again, if this import fails, <span class="application">Python</span> will raise an <tt class="errorcode">ImportError</tt>, which you catch.
                           </td>
</tr>
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.2.4"><img alt="4" border="0" height="12" src="../images/callouts/4.png" width="12"/></a>
</td>
<td align="left" valign="top">None of these platform-specific modules is available (which is possible, since <span class="application">Python</span> has been ported to a lot of different platforms), so you need to fall back on a default password input function (which is
                              defined elsewhere in the <tt class="filename">getpass</tt> module).  Notice what you're doing here: assigning the function <tt class="function">default_getpass</tt> to the variable <tt class="varname">getpass</tt>.  If you read the official <tt class="filename">getpass</tt> documentation, it tells you that the <tt class="filename">getpass</tt> module defines a <tt class="function">getpass</tt> function.  It does this by binding <tt class="varname">getpass</tt> to the correct function for your platform.  Then when you call the <tt class="function">getpass</tt> function, you're really calling a platform-specific function that this code has set up for you.  You don't need to know or
                              care which platform your code is running on -- just call <tt class="function">getpass</tt>, and it will always do the right thing.
                           </td>
</tr>
<tr>
<td align="left" valign="top" width="12"><a href="#fileinfo.exceptions.2.5"><img alt="5" border="0" height="12" src="../images/callouts/5.png" width="12"/></a>
</td>
<td align="left" valign="top">A <tt class="literal">try...except</tt> block can have an <tt class="literal">else</tt> clause, like an <tt class="literal">if</tt> statement.  If no exception is raised during the <tt class="literal">try</tt> block, the <tt class="literal">else</tt> clause is executed afterwards.  In this case, that means that the <tt class="literal">from EasyDialogs import AskPassword</tt> import worked, so you should bind <tt class="varname">getpass</tt> to the <tt class="function">AskPassword</tt> function.  Each of the other <tt class="literal">try...except</tt> blocks has similar <tt class="literal">else</tt> clauses to bind <tt class="varname">getpass</tt> to the appropriate function when you find an <tt class="literal">import</tt> that works.
                           </td>
</tr>
</table>
</div>
</div>
<div class="furtherreading">
<h3>Further Reading on Exception Handling</h3>
<ul>
<li><a href="http://www.python.org/doc/current/tut/tut.html"><i class="citetitle"><span class="application">Python</span> Tutorial</i></a> discusses <a href="http://www.python.org/doc/current/tut/node10.html#SECTION0010400000000000000000">defining and raising your own exceptions, and handling multiple exceptions at once</a>.
                     </li>
<li><a href="http://www.python.org/doc/current/lib/"><i class="citetitle"><span class="application">Python</span> Library Reference</i></a> summarizes <a href="http://www.python.org/doc/current/lib/module-exceptions.html">all the built-in exceptions</a>.
                     </li>
<li><a href="http://www.python.org/doc/current/lib/"><i class="citetitle"><span class="application">Python</span> Library Reference</i></a> documents the <a href="http://www.python.org/doc/current/lib/module-getpass.html">getpass</a> module.
                     </li>
<li><a href="http://www.python.org/doc/current/lib/"><i class="citetitle"><span class="application">Python</span> Library Reference</i></a> documents the <a href="http://www.python.org/doc/current/lib/module-traceback.html"><tt class="filename">traceback</tt> module</a>, which provides low-level access to exception attributes after an exception is raised.
                     </li>
<li><a href="http://www.python.org/doc/current/ref/"><i class="citetitle"><span class="application">Python</span> Reference Manual</i></a> discusses the inner workings of the <a href="http://www.python.org/doc/current/ref/try.html"><tt class="literal">try...except</tt> block</a>.
                     </li>
</ul>
</div>
</div>
</div>

            </div>
        </div>
        <div class="row">
            <div class="small-12 columns">
        
    <table border="0" cellpadding="0" cellspacing="0" class="Footer" summary="" width="100%">
<tr>
<td align="left" width="35%"><br><a class="NavigationArrow" href="../object_oriented_framework/summary.html">&lt;&lt;&#160;Summary</a></br></td>
<td align="center" width="30%"><br>&#160;<span class="divider">|</span>&#160;<span class="thispage">1</span> <span class="divider">|</span> <a href="file_objects.html" title="6.2.&#160;Working with File Objects">2</a> <span class="divider">|</span> <a href="for_loops.html" title="6.3.&#160;Iterating with for Loops">3</a> <span class="divider">|</span> <a href="more_on_modules.html" title="6.4.&#160;Using sys.modules">4</a> <span class="divider">|</span> <a href="os_module.html" title="6.5.&#160;Working with Directories">5</a> <span class="divider">|</span> <a href="all_together.html" title="6.6.&#160;Putting It All Together">6</a> <span class="divider">|</span> <a href="summary.html" title="6.7.&#160;Summary">7</a>&#160;<span class="divider">|</span>&#160;
            </br></td>
<td align="right" width="35%"><br><a class="NavigationArrow" href="file_objects.html">Working with File Objects&#160;&gt;&gt;</a></br></td>
</tr>
<tr>
<td colspan="3"><br/></td>
</tr>
</table>

        <!-- Footer -->
<div class="Footer">
    <p><a href="https://github.com/pcsforeducation/diveintopython/issues?sort=created&state=open">Report issues<img style="margin-left:10px" height="40" src="../images/logo_github.png"></a></p>
    <p class="copyright">Copyright &copy; 2000, 2001, 2002, 2003, 2004 <a href="mailto:josh@servercobra.com">Mark Pilgrim</a>. Maintained by Josh Gachnang.</p>
</div>
<script type="text/javascript"><!--
google_ad_client = "ca-pub-1879314568120827";
/* DiveIntoPython */
google_ad_slot = "8002881514";
google_ad_width = 468;
google_ad_height = 60;
//-->
</script>
<script type="text/javascript"
src="http://pagead2.googlesyndication.com/pagead/show_ads.js">
</script>
<!-- End Footer -->
            </div>
        </div>
        <!-- Google Analytics -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-9740779-18']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
<!-- End Google Analytics -->
        <script src="../js/jquery.js"></script>
        <script src="../js/foundation.min.js"></script>
        <script>
          $(document).foundation();
        </script>
    </body>
</html>